contribs/src/main/scala/io/delta/storage/BaseExternalLogStore.scala

```diff
@@ -0,0 +1,309 @@
+ * limitations under the License.
+ */
+
+package io.delta.storage
+
+import java.io.FileNotFoundException
+import java.net.URI
+import java.nio.charset.StandardCharsets.UTF_8
+import java.nio.file.FileSystemException
+
+import com.google.common.io.CountingOutputStream
+import org.apache.commons.io.IOUtils
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.delta.metering.DeltaLogging
+import org.apache.spark.sql.delta.util.FileNames
+import org.apache.spark.sql.delta.storage.HadoopFileSystemLogStore
+
+abstract class BaseExternalLogStore(
+    sparkConf: SparkConf,
+    hadoopConf: Configuration
+) extends HadoopFileSystemLogStore(sparkConf, hadoopConf)
+    with DeltaLogging {
+
+  /**
+   * Copies file within filesystem
+   * @param fs reference to [[FileSystem]]
+   * @param src path to source file
+   * @param dst path to destination file
+   */
+  private def copyFile(fs: FileSystem, src: Path, dst: Path): Unit = {
+    logDebug(s"copy file: $src -> $dst")
+    val input_stream = fs.open(src)
+    val output_stream = fs.create(dst, true)
+    try {
+      IOUtils.copy(input_stream, output_stream)
+      output_stream.close()
+    } finally {
+      input_stream.close()
+    }
+  }
+
+  private def resolvePath(
+      path: Path,
+      hadoopConf: Configuration
+  ): (FileSystem, Path) = {
+    val fs = path.getFileSystem(getHadoopConfiguration)
+    val resolvedPath = stripUserInfo(fs.makeQualified(path))
+    (fs, resolvedPath)
+  }
+
+  private def stripUserInfo(path: Path): Path = {
+    val uri = path.toUri
+    val newUri = new URI(
+      uri.getScheme,
+      null,
+      uri.getHost,
+      uri.getPort,
+      uri.getPath,
+      uri.getQuery,
+      uri.getFragment
+    )
+    new Path(newUri)
+  }
+
+  /**
+   * Method for assuring consistency on filesystem according to the external cache.
+   * Method tries to rewrite TransactionLog entry from temporary path if it does not exists.
+   * Method returns completed [[LogEntry]]
+   */
+
+  private def fixDeltaLog(fs: FileSystem, entry: LogEntry): Unit = {
+    if (entry.complete) {
+      return
+    }
+    var retry = 0;
+    var copied = false;
+
+    while (true) {
+      logDebug(
+        s"${if (retry > 0) "re" else ""}trying to fix: ${entry.fileName}"
+      )
+      try {
+        if (!copied && !fs.exists(entry.absoluteJsonPath)) {
+          fixDeltaLogCopyTempFile(fs, entry.absoluteTempPath, entry.absoluteJsonPath);
+          copied = true;
+        }
+        fixDeltaLogPutCompleteDbEntry(entry);
+        logInfo(s"fixed ${entry.fileName}")
+        return
+      } catch {
+        case e: Throwable =>
+          logInfo(s"${e.getClass.getName}: $e")
+          if (retry >= 3) throw e
+      }
+      retry += 1;
+    }
+  }
+
+  protected def fixDeltaLogCopyTempFile(fs: FileSystem, src: Path, dst: Path): Unit = {
```

Comment:
you can group together all 4 of these helper methods and then document above saying that we extract logic into a separate method so that we can more easily inject and test for failures